boot-qemu.py

```diff
@@ -0,0 +1,247 @@
+        if not self._qemu_arch:
+            raise RuntimeError('No QEMU architecture set?')
+        qemu_bin = f"qemu-system-{self._qemu_arch}"
+        if not (qemu_path := shutil.which(qemu_bin)):
+            raise RuntimeError(
+                f'{qemu_bin} could not be found on your system?')
+        self._qemu_path = Path(qemu_path)
+
+        # Locate kernel if it was not specified
+        if self.kernel:
+            if not self.kernel_dir:
+                self.kernel_dir = self.kernel.parent
+        else:
+            if not self.kernel_dir:
+                raise RuntimeError(
+                    'No kernel image or kernel build folder specified?')
+            if not self._default_kernel_path:
+                raise RuntimeError('No default kernel path specified?')
+
+            possible_kernel_locations = {
+                Path(self.kernel_dir,
+                     self._default_kernel_path),  # default (kbuild)
+                Path(self.kernel_dir,
+                     self._default_kernel_path.name),  # tuxmake
+            }
+            for loc in possible_kernel_locations:
+                if loc.exists():
+                    self.kernel = loc
+                    break
+            if not self.kernel:
+                possible_locations = "', '".join(
+                    str(path) for path in possible_kernel_locations)
+                raise FileNotFoundError(
+                    f"{self._default_kernel_path.name} could not be found at possible locations ('{possible_locations}')",
+                )
+
+        # Kernel options
+        if self.cmdline:
+            self._qemu_args += ['-append', ' '.join(self.cmdline)]
+        self._qemu_args += ['-kernel', self.kernel]
+        self._qemu_args += ['-initrd', self._prepare_initrd()]
+
+        # KVM
+        if self.use_kvm:
+            if not self.smp:
+                self.smp = self._get_default_smp_value()
+            self._qemu_args += ['-cpu', self._kvm_cpu, '-enable-kvm']
+
+        # Machine specs
+        self._qemu_args += ['-m', self._ram]
+        if self.smp:
+            self._qemu_args += ['-smp', str(self.smp)]
+
+        # Show information about QEMU
+        utils.green(f"QEMU location: \033[0m{self._qemu_path.parent}")
+        utils.green(f"QEMU version: \033[0m{self._get_qemu_ver_string()}")
+
+        # Pretty print and run QEMU command
+        qemu_cmd = [qemu_path, *self._qemu_args]
+        print(f"$ {' '.join(shlex.quote(str(elem)) for elem in qemu_cmd)}")
+        subprocess.run(qemu_cmd, check=True)
+
+
+class X86QEMURunner(QEMURunner):
+
+    def __init__(self):
+        super().__init__()
+
+        self.cmdline += ['console=ttyS0', 'earlycon=uart8250,io,0x3f8']
+
+        self._default_kernel_path = Path('arch/x86/boot/bzImage')
+        self._initrd_arch = 'x86'
+        self._qemu_arch = 'i386'
+
+    def _can_use_kvm(self):
+        return platform.machine() == 'x86_64' and self._have_dev_kvm_access()
```

Comment:
Similar to b845151d1e2e9c1296e0aaa83a4cc4dae861c5e5, I think this can also be a member set by subclass constructors rather than a method?